packages/fxa-auth-server/lib/subhub.js

```diff
@@ -0,0 +1,315 @@
+      method: 'POST',
+      validate: {
+        params: {
+          uid: isA.string().required(),
+        },
+        payload: {
+          pmt_token: isA.string().required(),
+          plan_id: isA.string().required(),
+          email: isA.string().required(),
+        },
+        response: isA.alternatives(
+          isA.object({
+            sub_id: isA.string()
+          }),
+          isA.object({
+            message: isA.string()
+          })
+        )
+      }
+    },
+
+    cancelSubscription: {
+      path: '/customer/:uid/subscriptions/:sub_id',
+      method: 'DELETE',
+      validate: {
+        params: {
+          uid: isA.string().required(),
+          sub_id: isA.string().required(),
+        },
+        response: isA.alternatives(
+          isA.object({}),
+          isA.object({
+            message: isA.string()
+          })
+        )
+      }
+    },
+
+  });
+
+  const api = new SubHubAPI(
+    config.subhub.url,
+    {
+      headers: {
+        // TODO: update with subhub final auth
+        Authorization: `Bearer ${config.subhub.key}`
+      },
+      timeout: 15000
+    }
+  );
+
+  return {
+    isStubAPI: false,
+
+    async listPlans() {
+      return api.listPlans();
+    },
+
+    async listSubscriptions(uid) {
+      try {
+        return await api.listSubscriptions(uid);
+      } catch (err) {
+        if (err.statusCode === 404) {
+          log.error('subhub.listSubscriptions.1', { uid, err });
+          // TODO: update with subhub listSubscriptions error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
+          }
+        }
+        throw err;
+      }
+    },
+
+    async createSubscription(uid, pmt_token, plan_id, email) {
+      try {
+        return await api.createSubscription(uid, { pmt_token, plan_id, email });
+      } catch (err) {
+        if (err.statusCode === 400) {
+          log.error('subhub.createSubscription.1', { uid, pmt_token, plan_id, email, err });
+          // TODO: update with subhub createSubscription error response for invalid payment token
+          if (err.message === 'invalid payment token') {
+            throw error.rejectedSubscriptionPaymentToken(pmt_token);
+          }
+          // TODO: update with subhub createSubscription error response for invalid plan ID
+          if (err.message === 'invalid plan id') {
+            throw error.unknownSubscriptionPlan(plan_id);
+          }
+        }
+        throw err;
+      }
+    },
+
+    async cancelSubscription(uid, sub_id) {
+      try {
+        return await api.cancelSubscription(uid, sub_id);
+      } catch (err) {
+        if (err.statusCode === 400 || err.statusCode === 404) {
+          log.error('subhub.cancelSubscription.1', { uid, sub_id, err });
+          // TODO: update with subhub cancelSubscription error response for invalid uid
+          if (err.message === 'invalid uid') {
+            throw error.unknownAccount(uid);
```

Comment:
Similar to the comment above, it seems like the uid is known to *us* otherwise this would not be called, just not to subhub. Maybe this warrants a new error number to indicate "no known subscriptions".

Reply:
Sure, I can create a new error for this in particular. Can always tweak it later as subhub materializes